### PR TITLE
cli: add configuration option to use or not use host netns

### LINKED ISSUE #731: Make it optional to start VM in the host netns
Now we always create a netns for shim and hypervisor processes. As discussion in https://github.com/kata-containers/runtime/issues/113#issuecomment-403684825, we should provide a option to let VM run in the host netns.

### PR DESCRIPTION
**Pull request by caoruidong**
If `disable_new_netns` set to true, create VM and shim processes in the host netns

Fixes #731

Signed-off-by: Ruidong Cao <caoruidong@huawei.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @caoruidong 

- This will be useful for tracing the agent (https://github.com/kata-containers/kata-containers/issues/27), so thanks for raising! :smile:
- Your branch has conflicts so please rebase.
- Please could you add a unit test for this new option.
- The issues in your "Fixes" comment mentions disabling the ns for the shim too, but the code only applies to the hypervisor.

**Comment by jodh-intel** [comment]
Oh - also, this new option should be displayed in the output of `kata-env` ideally (remember to increase the value of `formatVersion` in `kata-env.go`).

**Comment by caoruidong** [comment]
Actually config.NetNSPath will influence both shim and VM processes. I have changed commit msg.

But this doesn't work for CNM model. Because we scan endpoints by default and try to attach them, this will break host networking. And libnetwork hook will get a error `failed to set gateway while updating gateway: network is unreachable`.
So just works with `docker run -it --net=none`

File: cli/network.go [hunk 1]
```diff
@@ -337,13 +337,21 @@ func hostNetworkingRequested(configNetNs string) (bool, error) {
 
 func setupNetworkNamespace(config *vc.NetworkConfig) error {
 	if config.NetNSPath == "" {
-		n, err := ns.NewNS()
+		var (
+			n   ns.NetNS
+			err error
+		)
+		if config.DisableNewNetNs {
+			n, err = ns.GetNS("/proc/1/ns/net")
```
Comment (bergwolf) [inline cli/network.go]: Please use `/proc/self/ns/net` instead.

Or we can simply not to setup network namespace when `DisableNewNetNs` is set, in which case we can avoid `setupNetworkNamespace(`) in kata cli and sandbox `createNetwork()`/`removeNetwork()`.

**Comment by WeiZhang555** [comment]
This can be treated as an advanced feature and can only be enabled by users who really understand what this option means,  I think if user want to use docker libnetwork, they should keep this flag disabled, and only enable it when they want to do custom operations for network.  Based on this, I suggest we make a conflict rule:

* `disable_new_netns` conflicts with  `enable_netmon`
* `disable_new_netns` conflicts with `internetworking_model=bridge` and `internetworking_model=macvtap`, instead we add a new network model `none`. `disable_new_netns` works only with  `internetworking_model=none`

According to rule, any violation should got an error.

**Comment by jodh-intel** [comment]
Sounds good to me but let's also get input from @sboeuf, @amshinde, @mcastelino.

File: cli/network.go (hunk 1 above)
Comment (caoruidong) [reply cli/network.go]: Great! I have tried ns.GetCurrentNs() but forgot "/proc/self/ns/net".
I think it's better to skip setupNetworkNamespace(). Let's hear from others.

**Comment by caoruidong** [comment]
@bergwolf I have tried it with docker. I think it doesn't break host network but gets an error `failed to set gateway while updating gateway: network is unreachable.` due to hook.
It works with ` --net=none`

**Comment by mcastelino** [comment]
@sboeuf @amshinde I am surprised that --net=host did not cause any issues. Did you add new code that disabled networking when --net=host is setup? I understand --net=none working.

**Comment by amshinde** [comment]
@mcastelino Yes, --net=host has been diasbled explicitly, explicitly showing an error that this is not supported.

@caoruidong Is this feature for your custom networking?

> `disable_new_netns` conflicts with  `enable_netmon`
> `disable_new_netns` conflicts with `internetworking_model=bridge` and `internetworking_model=macvtap`, instead we add a new network model `none`. `disable_new_netns` works only with  `internetworking_model=none`

I agree with the error scenarios @WeiZhang555 pointed out. In addition, I think It would be a good idea to skip running the hooks altogether  when `disable_new_netns` is true.

**Comment by sboeuf** [comment]
@caoruidong based on input from @WeiZhang555 and @bergwolf , I think we need to add some documentation as part of this PR.
Because this option should only be used by ppl who know what they're doing, please add both documentation in the code and into a `.md` file.

Last thing, I would greatly appreciate if you could explain clearly the use case for such thing, since this is going against what we've been doing in Kata, and unless we have a strong use case for it, I don't know if we want it.

**Comment by caoruidong** [comment]
@mcastelino It is documented in https://github.com/kata-containers/documentation/blob/master/Limitations.md#docker---net=host.
This is kind of custom use case, but originally comes from what we discuss in https://github.com/kata-containers/runtime/issues/113#issuecomment-403684825. Thanks to gnawux, he draws a picture https://github.com/kata-containers/runtime/issues/113#issuecomment-403687004.
I will add some documents after we reach a consensus.

**Comment by WeiZhang555** [comment]
> I think It would be a good idea to skip running the hooks altogether when disable_new_netns is true.

@amshinde I'm not sure if there will be any new scenario depending on hook support in addition to "docker libnetwork", so maybe it's not good idea to disable hooks.

**Comment by caoruidong** [comment]
This modification may be too aggressive. Another way is like runv. VM starts in the host net ns but shim in a standalone net ns. WDYT?

**Comment by raravena80** [comment]
Hi @caoruidong, I'm the  Kata herder this week 😄  Just checking, any updates on this? Thx!

**Comment by caoruidong** [comment]
@raravena80 I want to some input from @sboeuf @amshinde. 
> In previous runv's implementation, the qemu is in root net ns, so we can easily plug the tap to qemu and then add the tap to a (let's say) OVS bridge which can gives us a better performance. But I think this can not apply to kata-runtime use case due to netns isolation, And the veth and additional bridge could bring network overhead and is not good.

We want to achieve better performance. The left one is runv's model. Key point is qemu. We can still leave shim in the net ns.
untitled drawing

**Comment by mcastelino** [comment]
@caoruidong the bridge has been eliminated in the later versions of Kata. We now support multiqueue macvtap that directly connects to the veth interface. 

At the same time I still like the runv tc model and we do plan to add support for it when we have some time in the future. 

/cc @amshinde @sboeuf I assume we have made macvtap as the default method in Kata?

**Comment by caoruidong** [comment]
@mcastelino Yes, default macvtap eliminated bridge. Though it still needs a veth pair. We may want to bypass veth pair too.

**Comment by amshinde** [comment]
@WeiZhang555 I was just being cautious so that we dont cause any side effects in the host network namespace. After all we don't know if support for adding custom hooks will be added to docker.
I dont want to add to  the bloat of configuration options, but maybe having a config option to run the hooks or not makes sense.
If not at the very least, we need to document `disable_new_netns` properly.

**Comment by WeiZhang555** [comment]
@mcastelino 
> @caoruidong the bridge has been eliminated in the later versions of Kata. We now support multiqueue macvtap that directly connects to the veth interface.
> At the same time I still like the runv tc model and we do plan to add support for it when we have some time in the future.

`veth` has quite low performance, I believe macvtap can't help much since it's backed by veth (correct me if I'm wrong). Let's provide the configuration option and let customer choose what they want by themselves.

File: virtcontainers/network.go [hunk 2]
```diff
@@ -995,6 +1007,47 @@ func bridgeNetworkPair(netPair *NetworkInterfacePair, numCPUs uint32, disableVho
 	return nil
 }
 
+func tuntapNetworkPair(endpoint *VirtualEndpoint, numCPUs uint32, disableVhostNet bool) error {
```
Comment (sboeuf) [inline virtcontainers/network.go]: Why are we doing things differently than `tapNetworkPair()`? If you're trying to tie an interface to the VM, you need this interface to be connected to the TAP interface being created, right?
I don't get why the model implemented by `tapNetworkPair()`, and relying on MACVTAP cannot be used for this use case?

File: cli/config.go [hunk 3]
```diff
@@ -577,9 +578,7 @@ func loadConfiguration(configPath string, ignoreLogging bool) (resolvedConfigPat
 		kataLog.Logger.Level = originalLoggerLevel
 	}
 
-	if tomlConf.Runtime.Tracing {
-		tracing = true
-	}
+	tracing = tomlConf.Runtime.Tracing
```
Comment (sboeuf) [inline cli/config.go]: you're modifying something that's not really part of this commit.

File: cli/config.go [hunk 4]
```diff
@@ -619,6 +623,18 @@ func loadConfiguration(configPath string, ignoreLogging bool) (resolvedConfigPat
 	return resolved, config, nil
 }
 
+func checkNetNsConfig(config oci.RuntimeConfig) error {
```
Comment (sboeuf) [inline cli/config.go]: I like this function :)
Please add a comment explaining what it does as this is an important sanity check of the configuration file.

File: cli/config.go [hunk 5]
```diff
@@ -619,6 +623,18 @@ func loadConfiguration(configPath string, ignoreLogging bool) (resolvedConfigPat
...
+	if config.DisableNewNetNs {
+		if config.NetmonConfig.Enable {
+			return fmt.Errorf("config disable_new_netns is conflict with enable_netmon")
```
Comment (sboeuf) [inline cli/config.go]: s/`is conflict with`/`conflicts with`/

File: cli/config/configuration.toml.in [hunk 6]
```diff
@@ -220,3 +220,8 @@ internetworking_model="@DEFNETWORKMODEL@"
 # (See https://www.jaegertracing.io/docs/getting-started).
 # (default: disabled)
 #enable_tracing = true
+
+# If enabled, the runtime will not create a net namespace for shim and
+# hypervisor process.
```
Comment (sboeuf) [inline cli/config/configuration.toml.in]:
```toml
# If enabled, the runtime will not create a network namespace for
# shim and hypervisor processes.
```

File: cli/network.go [hunk 7]
```diff
@@ -336,6 +336,10 @@ func hostNetworkingRequested(configNetNs string) (bool, error) {
 }
 
 func setupNetworkNamespace(config *vc.NetworkConfig) error {
+	if config.DisableNewNetNs {
+		return nil
```
Comment (sboeuf) [inline cli/network.go]: It'd be worth adding some log traces here to confirm that we're running without any netns!

File: cli/config.go (hunk 3 above)
Comment (caoruidong) [reply cli/config.go]: Because its cyclomatic complexity reaches the top, I must reduce one to add an "if" for netns config.

File: virtcontainers/network.go (hunk 2 above)
Comment (caoruidong) [reply virtcontainers/network.go]: We just create a tap device but no veth pair. So I create a new function. Maybe function name is confusing. It can reuse the tapNetworkPair() with some refactoring.

**Comment by caoruidong** [comment]
@sboeuf Thanks for your reviewing. I'll add some comments and documentation. Put .md files in this repo first.

File: cli/config/Config-Clarification.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+This option will not create a standalone network namespace for hypervisor (like qemu) and shim processes. They will
+live in the host network namespace.
+
+- This option may have some potential impacts to your host. It should only be used by people who know what they're doing.
+
+- `disable_new_netns` conflicts with `enable_netmon`
+
+- `disable_new_netns` conflicts with `internetworking_model=bridged` and `internetworking_model=macvtap`. It works only
+ with `internetworking_model=none`. This model is used when you want better performance.
```
Comment (WeiZhang555) [inline cli/config/Config-Clarification.md]: > This model is used when you want better performance.

This description is not good, it seems to indicate that our default network model has poor performance. I think it's better to say something like "This model is used when you want to do more custom works on networks".

I'll leave this to docs team

File: cli/config/Config-Clarification.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ 
+- If you are using docker, `disable_new_netns` only works with `docker run --net=none`
+
+# internetworking_model=none
+
+This model will only create a tap device. By cooperating with `disable_new_nets`, the tap device is in the host network
```
Comment (WeiZhang555) [inline cli/config/Config-Clarification.md]: s/disable_new_nets/disable_new_netns/

File: virtcontainers/network.go [hunk 10]
```diff
@@ -247,21 +255,21 @@ func (endpoint *VirtualEndpoint) Attach(h hypervisor) error {
 func (endpoint *VirtualEndpoint) Detach(netNsCreated bool, netNsPath string) error {
 	// The network namespace would have been deleted at this point
 	// if it has not been created by virtcontainers.
-	if !netNsCreated {
+	if !netNsCreated && netNsPath != "" {
```
Comment (WeiZhang555) [inline virtcontainers/network.go]: What's this modification for？

File: virtcontainers/network.go [hunk 11]
```diff
@@ -1013,6 +1025,47 @@ func bridgeNetworkPair(netPair *NetworkInterfacePair, numCPUs uint32, disableVho
...
+
+	if !disableVhostNet {
+		vhostFds, err := createVhostFds(int(numCPUs))
+		if err != nil {
+			return fmt.Errorf("Could not setup vhost fds %s : %s", netPair.VirtIface.Name, err)
+		}
+		netPair.VhostFds = vhostFds
+	}
+
+	vethLinkAttrs := &endpoint.EndpointProperties.Iface.LinkAttrs
```
Comment (WeiZhang555) [inline virtcontainers/network.go]: Wrong name? I suppose there shouldn't be any veth pair here.

**Comment by WeiZhang555** [comment]
Had a quick glance and most codes looks good :smile:

File: virtcontainers/network.go (hunk 10 above)
Comment (caoruidong) [reply virtcontainers/network.go]: This is because when use `disable_new_netns`, netNsCreated is false since we don't create a netns. Without this modification, `Detach` function returns without cleaning device. The tap device will leave in the host netns.

File: cli/config.go (hunk 4 above)
Comment (caoruidong) [reply cli/config.go]: Done

File: cli/config.go (hunk 5 above)
Comment (caoruidong) [reply cli/config.go]: Fixed.

File: cli/config.go [hunk 12]
```diff
@@ -619,6 +623,18 @@ func loadConfiguration(configPath string, ignoreLogging bool) (resolvedConfigPat
...
+		}
+		if config.InterNetworkModel != vc.NetXConnectNoneModel {
+			return fmt.Errorf("config disable_new_netns only works with none internetworking_model")
```
[the remark this answers is not in the document]
Comment (caoruidong) [reply cli/config.go]: "config disable_new_netns only works with none" Seems a little wired.

File: virtcontainers/network.go (hunk 11 above)
Comment (caoruidong) [reply virtcontainers/network.go]: Yes. Renamed.

File: cli/config.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (WeiZhang555) [reply cli/config.go]: No, I mean: 
`return fmt.Errorf("config disable_new_netns only works with 'none' internetworking_model")`

File: cli/network.go (hunk 7 above)
Comment (jodh-intel) [reply cli/network.go]: It might be overkill to add this to every log call, but due to the delicate nature of namespaces + golang, how about we log the namespace details at various points in the code [*]? We already log the pid so this will give more detail than a simple "changed namspace" log call:

```go
const nsPath = "/proc/self/ns"

func getNamespaces() (string, error) {
	files, err := ioutil.ReadDir(nsPath)
	if err != nil {
		return "", err
	}

	var ns []string

	for _, file := range files {
		filePath := path.Join(nsPath, file.Name())

		linkName, err := os.Readlink(filePath)
		if err != nil {
			return "", err
		}

		ns = append(ns, linkName)
	}

	sort.Sort(sort.StringSlice(ns))

	namespaces := strings.Join(ns, ",")

	return namespaces, nil
}
```

This will return something like:

```
cgroup:[4026531835],ipc:[4026531839],mnt:[4026531840],net:[4026532008],pid:[4026531836],pid:[4026531836],user:[4026531837],uts:[4026531838]
```

---
[*] - At runtime startup, whenever we change namespace, and in this function to show if we did or didn't switch ns.

File: cli/config.go [hunk 13]
```diff
@@ -640,6 +644,20 @@ func loadConfiguration(configPath string, ignoreLogging bool) (resolvedConfigPat
...
 
+// checkNetNsConfig performs sanity checks on disable_new_netns config.
+// Because it is an expert option and conflicts with some other common configs.
+func checkNetNsConfig(config oci.RuntimeConfig) error {
+	if config.DisableNewNetNs {
+		if config.NetmonConfig.Enable {
+			return fmt.Errorf("config disable_new_netns conflicts with enable_netmon")
+		}
+		if config.InterNetworkModel != vc.NetXConnectNoneModel {
+			return fmt.Errorf("config disable_new_netns only works with none")
```
[the remark this answers is not in the document]
Comment (caoruidong) [reply cli/config.go]: Done

File: cli/network.go (hunk 7 above)
Comment (caoruidong) [reply cli/network.go]: Yes, it's good to add some logs. But in this function we cannot get container id or sandbox id for log  fields.

**Comment by amshinde** [comment]
lgtm overall.
@sboeuf PTAL as well.

File: virtcontainers/network.go [hunk 14]
```diff
@@ -1267,6 +1278,47 @@ func bridgeNetworkPair(endpoint Endpoint, numCPUs uint32, disableVhostNet bool)
...
 }
 
+func tapNetwork(endpoint Endpoint, numCPUs uint32, disableVhostNet bool) error {
+	netHandle, err := netlink.NewHandle()
+	if err != nil {
+		return err
+	}
+	defer netHandle.Delete()
+
+	netPair := endpoint.NetworkPair()
```
[the remark this answers is not in the document]
Comment (caoruidong) [reply virtcontainers/network.go]: Yes, it is much clear to use another type. I will update this.

File: cli/config.go [hunk 15]
```diff
@@ -640,6 +644,20 @@ func loadConfiguration(configPath string, ignoreLogging bool) (resolvedConfigPat
...
 }
 
+// checkNetNsConfig performs sanity checks on disable_new_netns config.
+// Because it is an expert option and conflicts with some other common configs.
+func checkNetNsConfig(config oci.RuntimeConfig) error {
+	if config.DisableNewNetNs {
+		if config.NetmonConfig.Enable {
+			return fmt.Errorf("config disable_new_netns conflicts with enable_netmon")
+		}
+		if config.InterNetworkModel != vc.NetXConnectNoneModel {
```
Comment (bergwolf) [inline cli/config.go]: What about `NetXConnectEnlightenedModel`? Do we require it to always run inside a new netns?

File: virtcontainers/network.go [hunk 16]
```diff
@@ -1005,6 +1009,12 @@ func xconnectVMNetwork(endpoint Endpoint, connect bool, numCPUs uint32, disableV
 		return untapNetworkPair(endpoint)
 	case NetXConnectEnlightenedModel:
 		return fmt.Errorf("Unsupported networking model")
+	case NetXConnectNoneModel:
+		netPair.NetInterworkingModel = NetXConnectNoneModel
```
Comment (bergwolf) [inline virtcontainers/network.go]: Why setting `netPair.NetInterworkingModel` to its own value right after switching based on it?

File: cli/config.go (hunk 15 above)
Comment (caoruidong) [reply cli/config.go]: Since it is not implemented, I think it should be in a new netns like other models.

File: virtcontainers/network.go (hunk 16 above)
Comment (caoruidong) [reply virtcontainers/network.go]: Just follow other cases. These codes have not been changed since they are imported. I will delete these useless code.

**Comment by jodh-intel** [comment]
If I set the following in `configuration.toml`...

```
internetworking_model = "none"
disable_new_netns = true
```

... I get errors:

```
$ docker run -it --runtime=kata-runtime busybox sh
docker: Error response from daemon: OCI runtime create failed: exit status 1: stdout: , stderr: time="2018-10-09T09:52:06+01:00" level=fatal msg="failed to add interface veth35254a4 to sandbox: error setting interface \"veth35254a4\" IP to 172.17.0.2/16: cannot program address 172.17.0.2/16 in sandbox interface because it conflicts with existing route {Ifindex: 3 Dst: 172.17.0.0/16 Src: 172.17.0.1 Gw: <nil> Flags: [] Table: 254}": unknown.
```

**Comment by jodh-intel** [comment]
@caoruidong - thanks - I'd missed that comment in `configuration.toml`! ;) I guess there is no way to error out more cleanly if users don't specify this as `--net=none` doesn't affect `config.json`?

**Comment by caoruidong** [comment]
@jodh-intel Yes. I can check moreover or fix it before CI is repaired.

File: virtcontainers/network.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (sboeuf) [reply virtcontainers/network.go]: @caoruidong I agree with @amshinde here. It'd be better to create a new type, this would make things much clearer.

The argument of not being able to reuse the code defined in `VirtualEndpoint` does not sound valid to me, as we should be able to factorize (given the fact that we make it modular enough of course).

I think some refactoring is really necessary here as it's not appropriate to use VirtualEndpoint if your interface is not a `veth`.
But this should go further and as discussed with @amshinde, the type `VirtualEndpoint` should only be in charge of `veth`, which would make some code cleaner (we should not fallback onto `VirtualEndpoint` by default).

[the remark this answers is not in the document]
Comment (caoruidong) [reply virtcontainers/network.go]: Fine. I'll rewrite this part.

[the remark this answers is not in the document]
Comment (amshinde) [reply virtcontainers/network.go]: @sboeuf I have taken care of the default fallback case so that VirtualEndpoint will only be created in case of veth in my PR for network refactor:
https://github.com/kata-containers/runtime/pull/811/commits/ab15498bdfb1c24fc1e15bd4dabd919adc19f6b3

[the remark this answers is not in the document]
Comment (sboeuf) [reply virtcontainers/network.go]: Nice, thanks!

**Comment by caoruidong** [comment]
@sboeuf @amshinde Branch update. PTAL

File: virtcontainers/network.go [hunk 17]
```diff
@@ -448,13 +454,11 @@ func xconnectVMNetwork(endpoint Endpoint, connect bool, numCPUs uint32, disableV
 	}
 	switch netPair.NetInterworkingModel {
 	case NetXConnectBridgedModel:
-		netPair.NetInterworkingModel = NetXConnectBridgedModel
```
Comment (sboeuf) [inline virtcontainers/network.go]: Isn't it breaking a current behavior? Or did you remove it because it was not needed anymore?

File: virtcontainers/network.go [hunk 18]
```diff
@@ -448,13 +454,11 @@ func xconnectVMNetwork(endpoint Endpoint, connect bool, numCPUs uint32, disableV
...
 		if connect {
 			return bridgeNetworkPair(endpoint, numCPUs, disableVhostNet)
 		}
 		return unBridgeNetworkPair(endpoint)
 	case NetXConnectMacVtapModel:
-		netPair.NetInterworkingModel = NetXConnectMacVtapModel
```
Comment (sboeuf) [inline virtcontainers/network.go]: Same comment here.

File: virtcontainers/network.go [hunk 19]
```diff
@@ -832,6 +836,10 @@ func doNetNS(netNSPath string, cb func(ns.NetNS) error) error {
 	}
 	defer currentNS.Close()
 
+	if netNSPath == "" {
```
Comment (sboeuf) [inline virtcontainers/network.go]: Please add a comment here to make sure we don't miss the fact that if `netNSPath` is empty, the callback will be run in the current network namespace.

File: virtcontainers/network.go [hunk 20]
```diff
@@ -1110,11 +1118,14 @@ func createEndpoint(netInfo NetworkInfo, idx int, model NetInterworkingModel) (E
 		if socketPath != "" {
 			networkLogger().WithField("interface", netInfo.Iface.Name).Info("VhostUser network interface found")
 			endpoint, err = createVhostUserEndpoint(netInfo, socketPath)
+		} else if model == NetXConnectNoneModel {
```
Comment (sboeuf) [inline virtcontainers/network.go]: That's not right. We cannot rely on the internetworking model to decide we want to create a tap network endpoint.
Instead we should rely on:
```go
} else if netInfo.Iface.Type == TapEndpointType {
```
The internetworking model should only be used to understand how the created endpoint (`tap` in this case) is supposed to be connected to the host network.

File: virtcontainers/tap_endpoint.go [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+//
+
+package virtcontainers
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/containernetworking/plugins/pkg/ns"
+
```
Comment (sboeuf) [inline virtcontainers/tap_endpoint.go]: No need for blank lines between each imported package.

File: virtcontainers/tap_endpoint.go [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/kata-containers/runtime/virtcontainers/pkg/uuid"
+)
+
+// TapEndpoint represents just a tap endpoint
+type TapEndpoint struct {
+	ID                 string
+	NicName            string
+	TAPIface           NetworkInterface
+	VMFds              []*os.File
```
Comment (sboeuf) [inline virtcontainers/tap_endpoint.go]: Is this needed in case of `tap`? I thought only macvtap was requiring this.
/cc @amshinde

File: virtcontainers/tap_endpoint.go [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+	if err != nil {
+		return err
+	}
+	defer netHandle.Delete()
+
+	tapLink, fds, err := createLink(netHandle, endpoint.TAPIface.Name, &netlink.Tuntap{}, int(numCPUs))
+	if err != nil {
+		return fmt.Errorf("Could not create TAP interface: %s", err)
+	}
+	endpoint.VMFds = fds
```
Comment (sboeuf) [inline virtcontainers/tap_endpoint.go]: Is this needed?

File: virtcontainers/qemu.go [hunk 24]
```diff
@@ -833,67 +833,93 @@ func (q *qemu) hotplugVFIODevice(device *config.VFIODev, op operation) error {
...
+	devID := "virtio-" + id
+	if err = q.qmpMonitorCh.qmp.ExecuteNetPCIDeviceAdd(q.qmpMonitorCh.ctx, name, devID, hardAddr, addr, bridge.ID, int(q.config.NumVCPUs)); err != nil {
+		return pciAddr, err
+	}
+
+	return pciAddr, nil
 }
 
-func (q *qemu) hotplugNetDevice(drive *VethEndpoint, op operation) error {
+func (q *qemu) hotRemoveNetDevice(name, id string) error {
```
Comment (sboeuf) [inline virtcontainers/qemu.go]: This is a global comment about the factorization you're doing in `virtcontainers/qemu.go`. Please create a first commit introducing the factorization independently from the addition of `TapEndpoint`, and then you can add `TapEndpoint` with a following commit. This way, it's clearly defined what's done because of the factorization vs what's done by the addition of `TapEndpoint`.

File: virtcontainers/network.go (hunk 17 above)
Comment (caoruidong) [reply virtcontainers/network.go]: See https://github.com/kata-containers/runtime/pull/733#discussion_r223558592. IMO it is a redundance

Comment (sboeuf) [reply virtcontainers/network.go]: Looks fair to me (to remove it I mean), just please make sure there's not some weird behavior related to it that we're missing.

File: virtcontainers/tap_endpoint.go (hunk 22 above)
Comment (caoruidong) [reply virtcontainers/tap_endpoint.go]: When using qemu command line/qmp to insert a nic, We can use both name and fds. With this filed, tap/veth/macvtap can be hotpluged by the same way. Actually veth endpoint has this field too.

File: virtcontainers/network.go (hunk 20 above)
Comment (caoruidong) [reply virtcontainers/network.go]: This reminds me one thing. In hotplug network https://github.com/kata-containers/runtime/blob/d00742f43f3fb1177e74d864e47a1bae697c2d73/virtcontainers/sandbox.go#L1075-L1099, `netInfo.Iface.Type` is always "". So actually hotplug doesn't work now. You mean we set type there?

File: virtcontainers/network.go (hunk 19 above)
Comment (caoruidong) [reply virtcontainers/network.go]: Done

File: virtcontainers/qemu.go (hunk 24 above)
Comment (caoruidong) [reply virtcontainers/qemu.go]: Done

**Comment by caoruidong** [comment]
```
=== RUN   TestDeleteContainer
time="2018-10-17T09:43:33Z" level=error msg="trace called before context set" sandbox=7f49d00d-1995-4156-8c79-5f5ab24ce138 source=virtcontainers subsystem=sandbox type=bug
time="2018-10-17T09:43:33Z" level=error msg="trace called before context set" sandbox=7f49d00d-1995-4156-8c79-5f5ab24ce138 source=virtcontainers subsystem=sandbox type=bug
time="2018-10-17T09:43:33Z" level=error msg="trace called before context set" sandbox=7f49d00d-1995-4156-8c79-5f5ab24ce138 source=virtcontainers subsystem=sandbox type=bug
time="2018-10-17T09:43:33Z" level=error msg="failed to start proxy" error="Invalid proxy parameters {id:foobar path: agentURL: consoleURL: logger:0xc4200c9a40 debug:false}" params="{foobar    0xc4200c9a40 false}" proxy type=invalidProxyType source=virtcontainers vm=foobar
FAIL	github.com/kata-containers/runtime/virtcontainers	1.737s
```
@jodh-intel Is this a real error? My failed tests all meet this error. I remember it can be fixed by re-trigger. But it doesn't work this time.

**Comment by jodh-intel** [comment]
@caoruidong - the "trace called before context set" error messages can be ignored as this is only test code. But the last error looks like a real one.

**Comment by caoruidong** [comment]
@jodh-intel But this error comes out in vmfactory process. I can't find where TestCreateContainer or TestDeleteContainer calls vmfactory.

File: cli/config/configuration.toml.in [hunk 25]
```diff
@@ -230,9 +230,23 @@ path = "@NETMONPATH@"
 #   - macvtap
 #     Used when the Container network interface can be bridged using
 #     macvtap.
+#
+#   - none
+#     Used when customize network. Only creates a tap device. No veth pair.
```
Comment (jodh-intel) [inline cli/config/configuration.toml.in]: I think it would be useful to add a link here to the following as that shows users visually more about how this will work:

- https://github.com/kata-containers/documentation/blob/master/architecture.md#networking

Comment (caoruidong) [reply cli/config/configuration.toml.in]: OK. I will add a PR in documentation repo.

**Comment by caoruidong** [comment]
@bergwolf Comments addressed in this PR.

**Comment by raravena80** [comment]
@caoruidong ping from your weekly Kata herder. Doing re-test?

**Comment by caoruidong** [comment]
@amshinde @sboeuf branch update

**Comment by WeiZhang555** [comment]
LGTM.

As we got 4 LGTMs, I'll merge this now.
The design has been discussed a lot so I think it's safe to merge, we can keep improving detail implementation later.
